README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# diceR <a href='https://alinetalhouk.github.io/diceR/'><img src='man/figures/logo.png' align="right" width="120" /></a>

<!-- badges: start -->

[![R-CMD-check](https://github.com/AlineTalhouk/diceR/workflows/R-CMD-check/badge.svg)](https://github.com/AlineTalhouk/diceR/actions)
[![Codecov test
coverage](https://codecov.io/gh/AlineTalhouk/diceR/branch/master/graph/badge.svg)](https://app.codecov.io/gh/AlineTalhouk/diceR?branch=master)
[![CRAN
status](https://www.r-pkg.org/badges/version/diceR)](https://CRAN.R-project.org/package=diceR)
[![CRAN RStudio mirror
downloads](https://cranlogs.r-pkg.org/badges/grand-total/diceR?color=orange)](https://r-pkg.org/pkg/diceR)
<!-- badges: end -->

## Overview

The goal of `diceR` is to provide a systematic framework for generating
diverse cluster ensembles in R. There are a lot of nuances in cluster
analysis to consider. We provide a process and a suite of functions and
tools to implement a systematic framework for cluster discovery, guiding
the user through the generation of a diverse clustering solutions from
data, ensemble formation, algorithm selection and the arrival at a final
consensus solution. We have additionally developed visual and analytical
validation tools to help with the assessment of the final result. We
implemented a wrapper function `dice()` that allows the user to easily
obtain results and assess them. Thus, the package is accessible to both
end user with limited statistical knowledge. Full access to the package
is available for informaticians and statisticians and the functions are
easily expanded. More details can be found in our companion paper
published at [BMC
Bioinformatics](https://doi.org/10.1186/s12859-017-1996-y).

## Installation

You can install `diceR` from CRAN with:

``` r
install.packages("diceR")
```

Or get the latest development version from GitHub:

``` r
# install.packages("devtools")
devtools::install_github("AlineTalhouk/diceR")
```

## Example

The following example shows how to use the main function of the package,
`dice()`. A data matrix `hgsc` contains a subset of gene expression
measurements of High Grade Serous Carcinoma Ovarian cancer patients from
the Cancer Genome Atlas publicly available datasets. Samples as rows,
features as columns. The function below runs the package through the
`dice()` function. We specify (a range of) `nk` clusters over `reps`
subsamples of the data containing 80% of the full samples. We also
specify the clustering `algorithms` to be used and the ensemble
functions used to aggregated them in `cons.funs`.

``` r
library(diceR)
data(hgsc)
obj <- dice(hgsc, nk = 4, reps = 5, algorithms = c("hc", "diana"),
            cons.funs = c("kmodes", "majority"))
```

The first few cluster assignments are shown below:

``` r
knitr::kable(head(obj$clusters))
```

|                     | kmodes | majority |
|:--------------------|-------:|---------:|
| TCGA.04.1331_PRO.C5 |      2 |        2 |
| TCGA.04.1332_MES.C1 |      2 |        2 |
| TCGA.04.1336_DIF.C4 |      4 |        2 |
| TCGA.04.1337_MES.C1 |      2 |        2 |
| TCGA.04.1338_MES.C1 |      2 |        2 |
| TCGA.04.1341_PRO.C5 |      2 |        2 |

You can also compare the base `algorithms` with the `cons.funs` using
internal evaluation indices:

``` r
knitr::kable(obj$indices$ii$`4`)
```

|                 | Algorithms      | calinski_harabasz |      dunn |      pbm |       tau |     gamma |   c_index | davies_bouldin | mcclain_rao |    sd_dis |  ray_turi |    g_plus | silhouette | s_dbw | Compactness | Connectivity |
|:----------------|:----------------|------------------:|----------:|---------:|----------:|----------:|----------:|---------------:|------------:|----------:|----------:|----------:|-----------:|------:|------------:|-------------:|
| HC_Euclidean    | HC_Euclidean    |          3.104106 | 0.2608547 | 59.73711 | 0.1994677 | 0.6349401 | 0.2844073 |       1.839182 |   0.8009149 | 0.1306062 | 1.4765665 | 0.0180141 |        NaN |   NaN |    24.83225 |     41.62183 |
| DIANA_Euclidean | DIANA_Euclidean |         53.647400 | 0.3348103 | 33.87817 | 0.4296601 | 0.6314448 | 0.1589442 |       2.824201 |   0.8051915 | 0.2119281 | 3.2978986 | 0.0853202 |  0.0692233 |   NaN |    21.93396 |    241.66310 |
| kmodes          | kmodes          |         55.336853 | 0.3396909 | 50.73168 | 0.4553051 | 0.6600223 | 0.1455632 |       2.010999 |   0.7974362 | 0.1180643 | 1.1528004 | 0.0808925 |  0.1250821 |   NaN |    21.90504 |    204.58333 |
| majority        | majority        |         19.373248 | 0.3544371 | 85.05173 | 0.3694313 | 0.6529652 | 0.2102487 |       1.622799 |   0.8019453 | 0.1108674 | 0.9200511 | 0.0555431 |  0.1884934 |   NaN |    23.85408 |     64.04921 |

## Pipeline

This figure is a visual schematic of the pipeline that `dice()`
implements.

![Ensemble Clustering pipeline.](man/figures/pipeline.png)

Please visit the
[overview](https://alinetalhouk.github.io/diceR/articles/overview.html "diceR overview")
page for more detail.


NEWS.md
# diceR 1.2.0

* Use `testthat::skip_if_not_installed()` to run tests conditionally when using packages in Suggests
* Use roxygen tag `@examplesIf` and `rlang::is_installed()` to run examples conditionally when using packages in Suggests

# diceR 1.1.0

* Reinstate `blockcluster` package as it is under active maintenance again

# diceR 1.0.4

* Suppress new names messages from transformed NMF data

* Flattened matrices include 4th dimension of clustering array

# diceR 1.0.3

* Add package logo using `hexSticker`

* Package `blockcluster` has been archived, remove from Suggests

# diceR 1.0.2

* Remove deprecated `context()` in tests

* Add `RColorBrewer` to Suggests because it is cross referenced in `?graphs`

* Add trailing slash for URLs in DESCRIPTION

* Remove `tibble` from Imports, no longer used

# diceR 1.0.1

* Suppress warnings when `clValid::connectivity()` is called regarding comparison with more than one class. Since R-4.0.0, a `matrix` object also inherits from class `array`

* In `algii_heatmap()`, the object `ii` already has row names passed from columns which are named vectors (issue also identified in #148, thanks @romainfrancois)

# diceR 1.0.0

## Decreased dependencies

The following steps were taken to minimize dependencies and ensure that `diceR` can still run on R 3.5:

* Removed `cli` and `RColorBrewer` from Imports

* Moved `apcluster`, `blockcluster`, `cluster`, `dbscan`, `e1071`, `kernlab`, and `kohonen` to `Suggests`, use their specific clustering algorithms conditionally. `mclust` needs to be in `Imports` because `mclust::mclustBIC()` needs to be imported

* Moved `sigclust` to `Suggests`, use within `sigclust()` conditionally

* Moved `progress` to `Suggests`, use within `consensus_cluster()` conditionally

* Moved `poLCA` to `Suggests`, use within `dice()` conditionally

* Moved `Rtsne` to `Suggests`, use within `prepare_data()` conditionally

* Removed old dependency `grDevices` from `Imports`

* Set minimum version to R (>= 3.5) for `klaR` dependency `questionr`

* In `ev_confmat()`, use `yardstick::conf_mat()` instead of `caret::confusionMatrix()`. `caret` has many dependencies, so best to avoid using it

* In `graph_heatmap()`, use `NMF::aheatmap()` instead of `gplots::heatmap.2()`. `gplots` depends on `caTools`, which now relies on R (>= 3.6)

* In `consensus_cluster()`, use `stringr::str_to_title()` instead of `Hmisc::capitalize()`. `Hmisc` depends on `latticeExtra`, which now relies on R (>= 3.6)

* In `graph_delta_area()`, use base solution instead of `flux::auc()`. `flux` also depends on `caTools`

* In `prepare_data()`, use own implementation of `quantable::robustscale()` with all of the former function's defaults. `quantable` also depends on `caTools`

* Specify Bioconductor installation on Travis and AppVeyor since `NMF` now Imports `Biobase`

## Minor improvements and bug fixes

* Remove `suppressWarnings(RNGversion("3.5.0"))` after updating R version

* Run `LCA()` unit test on imputed clustering object

* Remove internal validity measures with any `Inf` entries for `consensus_reweigh()`

* Use a cleaner, more robust method of removing `Rplots.pdf` after running `test-graphs.R`

* Ensure column binding with `purrr::map_dfc()` in `consensus_rank()`

* Replaced `dplyr::bind_cols()` with `purrr::flatten_dfc()` to suppress warning "Outer names are only allowed for unnamed scalar" in `get_cdf()`

* update roxygen and docs

# diceR 0.6.0

* Remove deprecated `dplyr` functions and use `.data` pronoun

* k-means clustering should not support distance matrices as input (@jerryji1993, #139)

* Add LCA as a consensus function (@philstraforelli, #137)

# diceR 0.5.2

* Fix `length > 1 in coercion to logical` error in `consensus_evaluate()` due to comparisons using `||` operator

* Add `suppressWarnings(RNGversion("3.5.0"))` before call to `set.seed()` in examples, tests, and vignette to use old RNG sampling

* Use `.covrignore` to exclude `zzz.R` from being considered in code coverage

* Use `dplyr` version >= 0.7.5 to ensure `bind_rows()` works

* Fixed bug where scaled matrix using the "robust" method in `prepare_data()` was nested in the `data` element (@AlineTalhouk, #134)

# diceR 0.5.1

* Add parameter `hc.method` in `dice` and `consensus_cluster` to pass to `method` parameter in `stats::hclust` (@JakeNel28, #130)

* Remove dependencies on `largeVis`: package will be archived

# diceR 0.5.0

* Revert back to using `NMF` since `NNLM` has been archived and `NMF` is back in active maintenance.

* Choose fuzzifier m in `cmeans` using Equation 5 from https://academic.oup.com/bioinformatics/article/26/22/2841/227572 (thanks @Asduveneck)

# diceR 0.4.0

* Replace all code that depended on `NMF` with `NNLM` and `pheatmap`: CRAN notified that `NMF` will be archived because of inactive maintenance

* Update `.yml` files default templates

# diceR 0.3.2

* Fix bug in `consensus_cluster()` when custom algorithms were excluded from output (thanks @phiala)

* Use markdown language for documentation

* Various performance improvements and code simplifications

# diceR 0.3.1

* Suppress success/fail message printout and fix input data to be matrix for block clustering

* Fix bug in `algii_heatmap()` when `k.method = "all"` in `dice()`

* Fix bug in calculating internal indices when data has categorical variables (thanks Kurt Salmela)

# diceR 0.3.0

* Updated object output names in `consensus_evaluate()`

* Fix unit test in `test-dice.R` for R-devel

* Add internal function: ranked algorithms vs internal validity indices heatmap graph

* Fix bugs in `graph_cdf()`, `graph_tracking()` when only one k selected

* Progress messages in `dice()`

* Fix bug in `consensus_evaluate()` when algorithm has `NA` for all PAC values

# diceR 0.2.0

* New dimension reduction methods: t-SNE, largeVis (@dustin21)


R/algorithms.R

# Name Constants ----------------------------------------------------------

NALG <- "nmf"
DALG <- c("hc", "diana", "pam")
OALG <- c("km", "ap", "sc", "gmm", "block", "som", "cmeans", "hdbscan")
ALG_NAMES <- c(NALG, DALG, OALG)


# NMF-Based ---------------------------------------------------------------

#' Nonnegative Matrix Factorization
#' Transpose since input for NMF::nmf uses rows as vars, cols as samples
#' @noRd
nmf <- function(x, k, method, seed) {
  NMF::predict(NMF::nmf(t(x), rank = k, method = method, seed = seed))
}

#' Transform to non-negative matrix by column-binding a negative replicate and
#' then coercing all negative values to 0
#' @noRd
nmf_transform <- function(x) {
  x %>%
    cbind(-.) %>%
    as.data.frame() %>%
    purrr::map_dfc(~ ifelse(.x < 0, 0, .x)) %>%
    suppressMessages()
}

# Distance-Based ----------------------------------------------------------

#' Hierarchical Clustering
#' @noRd
hc <- function(d, k, method = "average") {
  as.integer(stats::cutree(stats::hclust(d, method = method), k))
}

#' DIvisive ANAlysis Clustering
#' @noRd
diana <- function(d, k) {
  if (!requireNamespace("cluster", quietly = TRUE)) {
    stop("Package \"cluster\" is needed. Please install it.",
         call. = FALSE)
  } else {
    as.integer(stats::cutree(cluster::diana(d, diss = TRUE), k))
  }
}

#' Partitioning Around Medoids
#' @noRd
pam <- function(d, k) {
  if (!requireNamespace("cluster", quietly = TRUE)) {
    stop("Package \"cluster\" is needed. Please install it.",
         call. = FALSE)
  } else {
    as.integer(cluster::pam(d, k, cluster.only = TRUE))
  }
}


# All Other Algorithms ----------------------------------------------------

#' K-Means Clustering
#' @noRd
km <- function(x, k) {
  as.integer(stats::kmeans(x, k)$cluster)
}

#' Affinity Propagation
#' @noRd
ap <- function(x, k) {
  if (!requireNamespace("apcluster", quietly = TRUE)) {
    stop("Package \"apcluster\" is needed. Please install it.",
         call. = FALSE)
  } else {
    suppressWarnings(
      apcluster::apclusterK(apcluster::negDistMat, x, k, verbose = FALSE)@idx
    ) %>%
      dplyr::dense_rank() %>%
      purrr::when(length(.) > 0 ~ ., ~ NA)
  }
}

#' Spectral Clustering (Radial-Basis Kernel)
#' @noRd
sc <- function(x, k) {
  if (!requireNamespace("kernlab", quietly = TRUE)) {
    stop("Package \"kernlab\" is needed. Please install it.",
         call. = FALSE)
  } else {
    kernlab::specc(as.matrix(x), k, kernel = "rbfdot")@.Data
  }
}

#' Gaussian Mixture Model
#' @noRd
gmm <- function(x, k) {
  mclust::Mclust(x, k, verbose = FALSE)$classification
}

#' Block Clustering (Co-clustering)
#' @noRd
block <- function(x, k) {
  if (!requireNamespace("blockcluster", quietly = TRUE)) {
    stop("Package \"blockcluster\" is needed. Please install it.",
         call. = FALSE)
  } else {
    tryCatch(
      sink_output(
        blockcluster::cocluster(as.matrix(x), "continuous",
                                nbcocluster = c(k, k))@rowclass + 1
      ),
      error = function(e) return(NA)
    ) %>%
      purrr::when(length(.) > 0 ~ ., ~ NA)
  }
}

#' Self-Organizing Maps
#' @noRd
som <- function(x, k, xdim, ydim, rlen, alpha, method) {
  if (!requireNamespace("kohonen", quietly = TRUE)) {
    stop("Package \"kohonen\" is needed. Please install it.",
         call. = FALSE)
  } else {
    x %>%
      purrr::when(is.matrix(.) ~ ., ~ as.matrix(.)) %>%
      som_train(xdim = xdim, ydim = ydim, rlen = rlen, alpha = alpha) %>%
      som_cluster(k = k, method = method)
  }
}

#' Train the SOM, specifiy grid size and other optional parameters based on the
#' SOM Toolbox
#' @references
#'   http://www.cis.hut.fi/somtoolbox/package/docs2/som_topol_struct.html
#' @noRd
som_train <- function(x, xdim, ydim, rlen, alpha, topo = "hexagonal") {
  # Create SOM grid and map data into the grid
  neurons <- 5 * sqrt(nrow(x))
  eigenvalues <- eigen(stats::cor(x))$values
  eigenratio <- eigenvalues[1] / eigenvalues[2]
  xdim <- xdim %||% sqrt(neurons / eigenratio)
  ydim <- ydim %||% neurons / xdim
  grid <- kohonen::somgrid(xdim = xdim, ydim = ydim, topo = topo)
  kohonen::som(x, grid = grid, rlen = rlen, alpha = alpha)
}

#' Perform hc: cut tree into k groups and output cluster labels for original data
#' @noRd
som_cluster <- function(model, k, method) {
  # Get distance matrix, use hc to cluster the codebook vectors
  cl <- hc(stats::dist(kohonen::getCodes(model, 1)), k = k, method = method)
  pred <- stats::predict(model)$unit.classif
  cl[pred]
}

#' Fuzzy C-Means (using best m via validity/performance measures)
#'
#' Fuzzifier m is a function of N and D (Equation 5 from ref.). If centers
#' can't be calculated, use frequently used value of m = 2.
#'
#' @references https://academic.oup.com/bioinformatics/article/26/22/2841/227572
#' @noRd
cmeans <- function(x, k) {
  if (!requireNamespace("e1071", quietly = TRUE)) {
    stop("Package \"e1071\" is needed. Please install it.",
         call. = FALSE)
  } else {
    N <- nrow(x)
    D <- ncol(x)
    m <- 1 + (1418 / N + 22.05) * D ^ (-2) +
      (12.33 / N + 0.243) * D ^ (-0.0406 * log(N) - 0.1134)
    fuzzy <- e1071::cmeans(x = x, centers = k, m = m)
    if (length(fuzzy$cluster) == 0) {
      fuzzy <- e1071::cmeans(x = x, centers = k, m = 2)
    }
    fuzzy$cluster
  }
}

#' Hierarchical Density-Based Spatial Clustering of Applications with Noise
#' @noRd
hdbscan <- function(x, minPts) {
  if (!requireNamespace("dbscan", quietly = TRUE)) {
    stop("Package \"dbscan\" is needed. Please install it.",
         call. = FALSE)
  } else {
    dbscan::hdbscan(x = x, minPts = minPts)$cluster
  }
}

#' Summarize the proportion of outliers and number of clusters
#' remove from consensus array and assign as an attribute, if used
#' @noRd
hdbscan_summarize <- function(arr) {
  h.idx <- match("HDBSCAN", dimnames(arr)[[3]])
  h.obj <- arr[, , h.idx, ] %>%
    as.data.frame() %>%
    purrr::map(~ {
      c(prop_outlier = sum(.x == 0, na.rm = TRUE) / sum(!is.na(.x)),
        num_cluster = dplyr::n_distinct(!.x %in% c(NA, 0)))
    }) %>%
    purrr::transpose() %>%
    purrr::map(unlist)
  arr <- arr[, , -h.idx, , drop = FALSE]
  attr(arr, "hdbscan") <- h.obj
  arr
}
